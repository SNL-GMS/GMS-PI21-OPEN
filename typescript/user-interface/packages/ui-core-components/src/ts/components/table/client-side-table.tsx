/* eslint-disable react/destructuring-assignment */
import { UILogger } from '@gms/ui-util';
import type { RowDataTransaction } from 'ag-grid-community';
import type Immutable from 'immutable';
import includes from 'lodash/includes';
import isEqual from 'lodash/isEqual';
import * as React from 'react';

import type { CoreTableProps } from './core-table';
import { CoreTable } from './core-table';
import type {
  ColumnApi,
  ColumnDefinition,
  ColumnGroupDefinition,
  GridReadyEvent,
  Row,
  TableApi
} from './types';

const logger = UILogger.create('GMS_LOG_TABLE', process.env.GMS_LOG_TABLE);

/** The Table Props
 * The context object is passed to most of the callbacks used in the grid.
 * The purpose of the context object is to allow the client application to
 * pass details to custom callbacks such as the Cell Renderers and Cell
 * Editors.
 */
export interface ClientSideTableProps<RowDataType, ContextDataType>
  extends CoreTableProps<RowDataType, ContextDataType> {
  rowDataNeverUpdates?: boolean;
}

/**
 * Table component that wraps AgGrid React.
 */
export class ClientSideTable<RowDataType extends Row, ContextDataType> extends React.Component<
  ClientSideTableProps<RowDataType, ContextDataType>,
  unknown
> {
  private refreshClientSideRowModelTimeout;

  /** Default row buffer */
  private readonly rowBuffer: number = 10;

  private coreTableRef: CoreTable<RowDataType, ContextDataType>;

  // ******************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * React lifecycle `shouldComponentUpdate`.
   * Determines if the component should update based on the next props passed in.
   *
   * @param nextProps props for the axis of type YAxisProps
   *
   * @returns boolean
   */
  public shouldComponentUpdate(
    nextProps: ClientSideTableProps<RowDataType, ContextDataType>
  ): boolean {
    // the component should only update (render) if the props have changed
    const propsHaveChanged = !isEqual(nextProps, this.props);
    return propsHaveChanged;
  }

  /**
   * React lifecycle `componentDidUpdate`.
   * Called immediately after updating occurs. Not called for the initial render.
   *
   * @param prevProps the previous props
   * @param prevState the previous state
   */
  public componentDidUpdate(prevProps: ClientSideTableProps<RowDataType, ContextDataType>): void {
    this.updateRowData(prevProps.rowData, this.props.rowData);

    if (this.getTableApi()) {
      if (this.refreshClientSideRowModelTimeout) {
        clearTimeout(this.refreshClientSideRowModelTimeout);
        this.refreshClientSideRowModelTimeout = undefined;
      }
      this.refreshClientSideRowModelTimeout = setTimeout(() => {
        this.getTableApi().refreshClientSideRowModel();
      });
    }
  }

  /**
   * React lifecycle `render`.
   * Renders the component.
   */
  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    return (
      <CoreTable
        ref={ref => {
          this.coreTableRef = ref;
        }}
        rowBuffer={this.rowBuffer}
        // user passed in props
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...this.props}
        /*
         * override user passed in props
         */
        onGridReady={this.onGridReady}
      />
    );
  }

  // ******************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Destroys the component instance.
   */
  public readonly destroy = (): void => {
    if (this.getTableApi()) {
      this.getTableApi().setRowData([]);
    }
    if (this.coreTableRef) {
      this.coreTableRef.destroy();
    }
    this.coreTableRef = undefined;
  };

  /**
   * Returns the Table Api instance.
   */
  public readonly getTableApi = (): TableApi => this.coreTableRef?.getTableApi();

  /**
   * Returns the Column Api instance.
   */
  public readonly getColumnApi = (): ColumnApi => this.coreTableRef?.getColumnApi();

  /**
   * Returns the number of visible columns.
   */
  public readonly getNumberOfVisibleColumns = (): number =>
    this.coreTableRef?.getNumberOfVisibleColumns();

  /**
   * Sets a column as visible or not visible (show/hide).
   *
   * @param key the unique key for the column
   * @param visible true to show; false to hide
   */
  public readonly setColumnVisible = (key: string, visible: boolean): void => {
    if (this.coreTableRef) {
      this.coreTableRef.setColumnVisible(key, visible);
    }
  };

  /**
   * Sets the column definitions for the table.
   *
   * @param columnDefinitions the column definitions
   */
  public readonly setColumnDefinitions = (
    columnDefinitions: (
      | ColumnDefinition<RowDataType, ContextDataType, unknown, unknown, unknown>
      | ColumnGroupDefinition
    )[]
  ): void => this.coreTableRef?.setColumnDefinitions(columnDefinitions);

  // ******************************************
  // START PRIVATE METHODS
  // ******************************************

  /**
   * Event handler for the `onGridReady` event. This is fired once when the ag-grid is initially created.
   * Invokes the users' `onGridReady` callback when specified.
   *
   * @param event The grid ready even
   */
  private readonly onGridReady = (event: GridReadyEvent): void => {
    this.updateRowData([], this.props.rowData);

    if (this.props.onGridReady) {
      this.props.onGridReady(event);
    }
  };

  /**
   * Determine the `changes` between row data (creates a RowDataTransaction).
   *
   * @param previousRowData the previous or current row data
   * @param rowData the new row data
   */
  private readonly determineRowDataChanges = (
    previousRowData: RowDataType[],
    rowData: RowDataType[]
  ): RowDataTransaction => {
    const prevIds = previousRowData.map(previousRow => previousRow.id);
    const newIds = rowData.map(newRow => newRow.id);
    const addData = rowData.filter(newRow => !includes(prevIds, newRow.id));
    const removeData = previousRowData.filter(previousRow => !includes(newIds, previousRow.id));
    const updateData = this.props.rowDataNeverUpdates
      ? []
      : rowData
          .filter(newRow => includes(prevIds, newRow.id))
          .filter(
            newRow =>
              !isEqual(
                newRow,
                previousRowData.find(previousRow => newRow.id === previousRow.id)
              )
          );
    return {
      add: addData,
      remove: removeData,
      update: updateData
    };
  };

  /**
   * Update the row data for the table.
   *
   * @param previousRowData the previous or current row data
   * @param rowData the new row data
   */
  private readonly updateRowData = (
    previousRowData: RowDataType[],
    rowData: RowDataType[]
  ): void => {
    if (this.getTableApi()) {
      // update the row data if it has changed
      if ((rowData && rowData.length === 0) || !isEqual(rowData, previousRowData)) {
        if (this.props.debug) {
          logger.debug(`table applyTransactionAsync`);
        }

        const changes: RowDataTransaction = this.determineRowDataChanges(previousRowData, rowData);

        // batch add, remove, and update the rows that changed
        if (this.getTableApi()) {
          this.getTableApi().applyTransaction(changes);
        }

        // refresh the updated row nodes
        if (this.coreTableRef) {
          this.coreTableRef.refreshRowNodesByIds(
            changes.update.map((change: RowDataType) => change.id)
          );
        }
      }
    }
  };

  /**
   * Toggles visible columns
   *
   * @param columnsToDisplay enum of columnNames to display map
   */
  public updateVisibleColumns<T extends string>(columnsToDisplay: Immutable.Map<T, boolean>): void {
    if (this.coreTableRef) {
      this.coreTableRef.updateVisibleColumns<T>(columnsToDisplay);
    }
  }

  // ******************************************
  // END PRIVATE METHODS
  // ******************************************
}
