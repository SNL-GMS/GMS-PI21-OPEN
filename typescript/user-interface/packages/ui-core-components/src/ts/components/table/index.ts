export { ClientSideTable as Table } from './client-side-table';
export { InfiniteTable } from './infinite-table';
export { PercentBar } from './percent-bar';
export {
  getColumnPosition,
  TableCellRenderer,
  TableCellRendererFramework
} from './table-cell-renderer';
export * from './types';
export { Datasource, GetRowsParams } from './types/datasource';
export { useDatasource } from './use-datasource';
export * as AgGridCommunity from 'ag-grid-community';
export { AgGridReact } from 'ag-grid-react';
