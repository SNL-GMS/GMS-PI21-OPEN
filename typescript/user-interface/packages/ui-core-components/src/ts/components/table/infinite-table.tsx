/* eslint-disable react/destructuring-assignment */
import { UILogger } from '@gms/ui-util';
import type { IDatasource, IGetRowsParams } from 'ag-grid-community';
import type Immutable from 'immutable';
import merge from 'lodash/merge';
import * as React from 'react';

import type { CoreTableProps } from './core-table';
import { CoreTable } from './core-table';
import type {
  ColumnApi,
  ColumnDefinition,
  ColumnGroupDefinition,
  GridReadyEvent,
  Row,
  TableApi
} from './types';
import type { Datasource } from './types/datasource';

const logger = UILogger.create('GMS_LOG_TABLE', process.env.GMS_LOG_TABLE);

/** The Table Props */
export interface InfiniteTableProps<RowDataType, ContextDataType>
  extends CoreTableProps<RowDataType, ContextDataType> {
  rowCount?: number;
  datasource?: Datasource;
}

/**
 * Table component that wraps AgGrid React.
 */
export class InfiniteTable<RowDataType extends Row, ContextDataType> extends React.Component<
  InfiniteTableProps<RowDataType, ContextDataType>,
  unknown
> {
  /** Default row buffer */
  private readonly defaultRowBuffer: number = 2;

  /** Default cache block size (how many rows per block to request from the datasource) */
  private readonly defaultCacheBlockSize: number = 100;

  /** the number of (empty) rows to show by default */
  private readonly defaultInitialRowCount: number = 5;

  /** the number of milliseconds to debounce datasource getRows calls */
  private readonly defaultBlockDebounceMs: number = 60;

  /**
   * The default number of blocks to store in cache.
   * Block size is set by defaultCacheBlockSize or user defined props
   */
  private readonly defaultMaxBlocksInCache: number = 2;

  private readonly defaultCacheOverflowSize: number = 10;

  private coreTableRef: CoreTable<RowDataType, ContextDataType>;

  // ******************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * React lifecycle `componentDidUpdate`.
   * Called immediately after updating occurs. Not called for the initial render.
   *
   * @param prevProps the previous props
   * @param prevState the previous state
   */
  public componentDidUpdate(prevProps: InfiniteTableProps<RowDataType, ContextDataType>): void {
    if (this.getTableApi()) {
      const rowCount = this.props.rowData?.length ?? 0;
      const prevRowCount = prevProps.rowData?.length ?? 0;
      // TODO: create better test to see if it changed
      if (rowCount !== prevRowCount) {
        if (this.getTableApi()) {
          this.getTableApi().refreshInfiniteCache();
          this.getTableApi().setRowCount(rowCount, true);
        }
      }
    }
  }

  /**
   * React lifecycle `componentWillUnmount`.
   * Called immediately before a component is destroyed. Perform any necessary
   * cleanup in this method, such as canceled network requests,
   * or cleaning up any DOM elements created in componentDidMount.
   */
  public componentWillUnmount(): void {
    if (this.props.debug) {
      logger.debug(`table componentWillUnmount`);
    }
    this.destroy();
  }

  /**
   * React lifecycle `render`.
   * Renders the component.
   */
  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    return (
      <CoreTable
        ref={ref => {
          this.coreTableRef = ref;
        }}
        // Defaults infinite scroll props (may be overridden by user supplied props)
        cacheBlockSize={this.defaultCacheBlockSize}
        maxBlocksInCache={this.defaultMaxBlocksInCache}
        cacheOverflowSize={this.defaultCacheOverflowSize}
        infiniteInitialRowCount={this.defaultInitialRowCount}
        blockLoadDebounceMillis={this.defaultBlockDebounceMs}
        rowBuffer={this.defaultRowBuffer}
        // user passed in props
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...this.props}
        // * Overrides to user-supplied props

        // force the rowModelType to be infinite for this table
        rowModelType="infinite"
        /*
         * datasource is overridden from props so that it does not trigger a rerender that causes a flash.
         * datasource is set in onGridReady callback
         */
        datasource={null}
        // Merge in the tableApi and columnApi to the ag grid context
        context={merge(this.props.context, {
          tableApi: this.getTableApi(),
          columnApi: this.getColumnApi()
        })}
        // Add our own onGridReady to be called by the CoreTable
        onGridReady={this.onGridReady}
      />
    );
  }

  // ******************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Destroys the component instance.
   */
  public readonly destroy = (): void => {
    if (this.getTableApi()) {
      const tableApi = this.coreTableRef.getTableApi();
      const emptyDataSource = this.getEmptyDataSource();
      tableApi.setDatasource(emptyDataSource);
      tableApi.purgeInfiniteCache();
    }
    if (this.coreTableRef) {
      this.coreTableRef.destroy();
    }
    this.coreTableRef = undefined;
  };

  /**
   * Returns the Table Api instance.
   */
  public readonly getTableApi = (): TableApi => this.coreTableRef?.getTableApi();

  /**
   * Returns the Column Api instance.
   */
  public readonly getColumnApi = (): ColumnApi => this.coreTableRef?.getColumnApi();

  /**
   * Returns the number of visible columns.
   */
  public readonly getNumberOfVisibleColumns = (): number =>
    this.coreTableRef?.getNumberOfVisibleColumns();

  /**
   * Sets a column as visible or not visible (show/hide).
   *
   * @param key the unique key for the column
   * @param visible true to show; false to hide
   */
  public readonly setColumnVisible = (key: string, visible: boolean): void => {
    if (this.coreTableRef) {
      this.coreTableRef.setColumnVisible(key, visible);
    }
  };

  /**
   * Sets the column definitions for the table.
   *
   * @param columnDefinitions the column definitions
   */
  public readonly setColumnDefinitions = (
    columnDefinitions: (
      | ColumnDefinition<RowDataType, ContextDataType, unknown, unknown, unknown>
      | ColumnGroupDefinition
    )[]
  ): void => this.coreTableRef?.setColumnDefinitions(columnDefinitions);

  /**
   * Called from destroy, reset the datasource
   *
   * @returns an empty IDatasource
   */
  // eslint-disable-next-line class-methods-use-this
  public readonly getEmptyDataSource = (): IDatasource => {
    return {
      getRows(params: IGetRowsParams) {
        params.successCallback([], 0);
      }
    };
  };

  // ******************************************
  // START PRIVATE METHODS
  // ******************************************

  /**
   * Event handler for the `onGridReady` event. This is fired once when the ag-grid is initially created.
   * Invokes the users' `onGridReady` callback when specified.
   *
   * @param event The grid ready even
   */
  private readonly onGridReady = (event: GridReadyEvent): void => {
    if (this.props.datasource) {
      this.getTableApi().setDatasource(this.props.datasource);
    }

    if (this.props.onGridReady) {
      this.props.onGridReady(event);
    }
  };

  /**
   * Toggles visible columns
   *
   * @param columnsToDisplay enum of columnNames to display map
   */
  public updateVisibleColumns<T extends string>(columnsToDisplay: Immutable.Map<T, boolean>): void {
    if (this.coreTableRef) {
      this.coreTableRef.updateVisibleColumns<T>(columnsToDisplay);
    }
  }

  // ******************************************
  // END PRIVATE METHODS
  // ******************************************
}
