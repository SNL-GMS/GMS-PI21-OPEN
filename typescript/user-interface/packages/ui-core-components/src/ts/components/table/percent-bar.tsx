/* eslint-disable react/destructuring-assignment */
import uniqueId from 'lodash/uniqueId';
import React from 'react';

import type { PercentBarProps } from './types/percent-bar';

/**
 * A simple component for rendering a percent bar.
 */
export class PercentBar extends React.PureComponent<PercentBarProps> {
  private readonly id: string;

  public constructor(props: PercentBarProps) {
    super(props);
    this.id = uniqueId();
  }

  public render(): JSX.Element {
    return (
      <div
        className={`percent-bar${this.props.classNames ? ' ' : ''}${
          this.props.classNames ? this.props.classNames : ''
        }`}
        key={this.id}
        style={{
          width: `100%`,
          // Parent has overflow hidden, which cuts off the overflow,
          // so translation will only show the appropriate amount of the bar.
          // eslint-disable-next-line @typescript-eslint/no-magic-numbers
          transform: `translateX(-${100 - (this.props.percentage || 0)}%)`
        }}
      />
    );
  }
}
