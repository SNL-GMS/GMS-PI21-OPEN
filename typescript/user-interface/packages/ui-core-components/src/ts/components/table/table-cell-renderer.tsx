/* eslint-disable react/prop-types */
import * as StringUtils from '@gms/common-util';
import { classList, getDataAttributesFromProps } from '@gms/ui-util';
import * as React from 'react';

import type { CellRendererParams } from './types/cell-renderer';

/**
 * Props for generic table cell renderer
 */
export interface TableCellRendererProps {
  cellValueClassName?: string;
  children?: JSX.Element | JSX.Element[] | string | React.ReactNode;
  className?: string;
  heightCSS?: string;
  isDate?: boolean;
  isNumeric?: boolean;
  leftChild?: JSX.Element;
  shouldCenterText?: boolean;
  tooltipMsg?: string;
  value?: string;
}

/**
 * Cell renderer for table cells. Accepts classes,
 * data-attributes, and a flag to indicate if it is numeric
 */
// eslint-disable-next-line react/function-component-definition
const TableCellComponent: React.FunctionComponent<TableCellRendererProps> = (
  props: TableCellRendererProps
) => {
  const dataAttributes = getDataAttributesFromProps(props);
  const {
    cellValueClassName,
    children,
    className,
    heightCSS,
    isDate,
    isNumeric,
    leftChild,
    shouldCenterText,
    tooltipMsg,
    value
  } = props;
  return (
    <div
      style={{
        height: heightCSS || '36px',
        overflow: 'hidden'
      }}
      data-cy="table-cell"
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dataAttributes}
      className={`table-cell ${className ?? ''}`}
      title={tooltipMsg || value}
    >
      {leftChild}
      {value ? (
        <div
          className={classList(
            {
              'table-cell__value--date': isDate,
              'table-cell__value--numeric': isNumeric,
              'table-cell__value--center': shouldCenterText
            },
            `table-cell__value ${cellValueClassName ?? ''}`
          )}
          data-cy="table-cell__value"
        >
          <span>{value}</span>
        </div>
      ) : null}
      {children}
    </div>
  );
};

/**
 * Cell renderer for table cells. Accepts classes,
 * data-attributes, and a flag to indicate if it is numeric
 */
export const TableCellRenderer = React.memo(TableCellComponent);

/**
 * Generic column position method used to get the column position.
 *
 * @param props generic cell renderer props
 * @returns column position
 */
export function getColumnPosition<T>(
  props: CellRendererParams<T, unknown, unknown, unknown, unknown>
): number | 'first' | 'last' {
  const index = props.columnApi
    ?.getAllDisplayedColumns()
    ?.findIndex(c => c.getColId() === (props.colDef.colId ?? props.colDef.field));
  let retValue: number | 'first' | 'last' = index;
  if (index === 0) {
    retValue = 'first';
  } else if (index === (props.columnApi?.getAllDisplayedColumns()?.length ?? -Infinity) - 1) {
    retValue = 'last';
  }
  return retValue;
}

/**
 * Generic table cell renderer framework used for general renderer in column definitions
 *
 * @param props generic cell renderer props
 * @returns JSX.Element of what to render
 */
export function TableCellRendererFramework<T>(
  props: CellRendererParams<T, unknown, any, unknown, unknown>
): JSX.Element {
  const { valueFormatted, value, colDef } = props;
  const tooltipMsg =
    colDef && colDef.tooltipValueGetter
      ? colDef.tooltipValueGetter({ ...props, location: 'cell' })
      : undefined;
  return (
    <TableCellRenderer
      data-col-position={getColumnPosition<T>(props)}
      isDate={StringUtils.isDate(value?.toString())}
      isNumeric={StringUtils.isNumeric(value?.toString())}
      tooltipMsg={tooltipMsg ?? valueFormatted ?? value}
      value={valueFormatted ?? value}
    />
  );
}
