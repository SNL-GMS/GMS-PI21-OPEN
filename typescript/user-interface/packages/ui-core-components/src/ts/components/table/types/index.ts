export * from './cell-renderer';
export * from './column-api';
export * from './column-definition';
export * from './events';
export * from './percent-bar';
export * from './row-node';
export * from './table-api';
export * from './table-options';
export * from './tooltip-renderer';
export * from './types';
