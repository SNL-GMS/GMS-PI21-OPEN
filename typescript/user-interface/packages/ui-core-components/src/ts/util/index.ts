export * from './display-util';
export * from './non-ideal-state';
