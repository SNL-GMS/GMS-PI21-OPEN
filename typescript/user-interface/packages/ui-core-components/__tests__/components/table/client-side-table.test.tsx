/* eslint-disable react/jsx-props-no-spreading */
import * as Enzyme from 'enzyme';
import Immutable from 'immutable';
import * as React from 'react';

import type { ClientSideTableProps } from '../../../src/ts/components/table/client-side-table';
import { ClientSideTable } from '../../../src/ts/components/table/client-side-table';
import type { ColumnDefinition, Row } from '../../../src/ts/ui-core-components';
import { waitForComponentToPaint } from '../../util/test-util';

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

enum ColumnNames {
  test = 'test',
  test2 = 'test2'
}

interface RowType extends Row {
  versionId: number;
  color: string;
  category: string;
  type: string;
  startTime: number;
  endTime: number;
  Ids: number[];
  rationale: string;
}

const generateTableRows = () => {
  const rows: RowType[] = [
    {
      id: '1',
      versionId: 1,
      color: 'RED',
      category: 'normal',
      type: 'info',
      startTime: 1000,
      endTime: 2000,
      Ids: [1, 2],
      rationale: 'user'
    },
    {
      id: '2',
      versionId: 1,
      color: 'GREEN',
      category: 'normal',
      type: 'info',
      startTime: 2000,
      endTime: 3000,
      Ids: [1, 2],
      rationale: 'user'
    }
  ];
  rows[0]['first-in-table'] = true;
  return rows;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const columns: any[] = [
  {
    headerName: 'test',
    field: 'test',
    headerTooltip: 'test',
    width: 10
  },
  {
    headerName: 'test2',
    field: 'test2',
    headerTooltip: 'test2',
    width: 10,
    sort: 'asc'
  }
];

const columnDefs: ColumnDefinition<RowType, unknown, unknown, unknown, unknown>[] = [...columns];

describe('Client Side Table', () => {
  const clientSideTableProps: ClientSideTableProps<RowType, any> = {
    defaultColDef: undefined,
    onCellClicked: jest.fn(),
    onCellContextMenu: jest.fn(),
    onRowSelected: jest.fn(),
    columnDefs: columnDefs as any,
    rowData: generateTableRows()
  };

  const defaultColumnWidthPx = 200;
  const headerCellBlockClass = 'table-header-cell';

  const compareCellValues = (a: number, b: number): number => {
    if (a === undefined && b === undefined) {
      return 0;
    }
    if (a === undefined) {
      return -1;
    }
    if (b === undefined) {
      return 1;
    }
    return a - b;
  };

  const defaultColumnDefinition = {
    headerClass: [`${headerCellBlockClass}`, `${headerCellBlockClass}--neutral`],
    width: defaultColumnWidthPx,
    sortable: true,
    filter: true,
    disableStaticMarkupForHeaderComponentFramework: true,
    disableStaticMarkupForCellRendererFramework: true,
    comparator: compareCellValues
  };

  // eslint-disable-next-line react/jsx-props-no-spreading
  const mockClientSideTable = Enzyme.shallow(<ClientSideTable {...clientSideTableProps} />);
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('is exported', () => {
    expect(ClientSideTable).toBeDefined();
  });
  it('Renders', () => {
    expect(mockClientSideTable).toMatchSnapshot();
  });

  // TODO: fix when migrating off of Enzyme--passes locally but fails in pipeline
  it.skip('can update', async () => {
    const wrapper = Enzyme.mount(<ClientSideTable {...clientSideTableProps} />);
    const instance: ClientSideTable<any, any> = wrapper
      .find(ClientSideTable)
      .instance() as ClientSideTable<any, any>;
    const spy = jest.spyOn(instance, 'shouldComponentUpdate');
    const spy2 = jest.spyOn(instance, 'componentDidUpdate');
    const spy3 = jest.spyOn(instance, 'componentDidUpdate');
    const spy4 = jest.spyOn(instance, 'render');
    wrapper.setProps({ defaultColDef: defaultColumnDefinition });
    await waitForComponentToPaint(wrapper);

    expect(spy).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).toHaveBeenCalled();
    expect(spy4).toHaveBeenCalled();

    const tableAPI = instance.getTableApi();
    expect(tableAPI).toBeDefined();

    const columnAPI = instance.getColumnApi();
    expect(columnAPI).toBeDefined();

    const visibleColumns = instance.getNumberOfVisibleColumns();
    expect(visibleColumns).toBeDefined();

    let columnsMap = Immutable.Map(
      Object.values(ColumnNames)
        // all columns are visible by default
        .map(v => [v, true])
    );

    instance.updateVisibleColumns<ColumnNames>(columnsMap);
    expect(instance.getNumberOfVisibleColumns()).toEqual(2);

    columnsMap = columnsMap.set(ColumnNames.test, false);

    await waitForComponentToPaint(wrapper);

    instance.updateVisibleColumns<ColumnNames>(columnsMap);
    expect(instance.getNumberOfVisibleColumns()).toEqual(1);
  });
  it('will not clear refresh client side row model timeout if not needed', () => {
    const clearTimeoutSpy = jest.spyOn(global, 'clearTimeout');
    const setTimeoutSpy = jest.spyOn(global, 'setTimeout');
    const tableWrapper = Enzyme.mount(<ClientSideTable {...clientSideTableProps} />);
    const tableWrapperInstance: ClientSideTable<any, any> = tableWrapper
      .find(ClientSideTable)
      .instance() as ClientSideTable<any, any>;
    Reflect.set(
      tableWrapperInstance,
      'getTableApi',
      jest.fn(() => ({
        ensureIndexVisible: jest.fn(() => true),
        refreshInfiniteCache: jest.fn()
      }))
    );
    tableWrapper.update();
    tableWrapper.instance().forceUpdate();
    expect(setTimeoutSpy).toHaveBeenCalled();
    expect(clearTimeoutSpy).not.toHaveBeenCalled();
  });

  // TODO Unskip tests and fix
  it.skip('will clear refresh client side row model timeout if needed', () => {
    const clearTimeoutSpy = jest.spyOn(global, 'clearTimeout');
    const setTimeoutSpy = jest.spyOn(global, 'setTimeout');
    const tableWrapper = Enzyme.mount(<ClientSideTable {...clientSideTableProps} />);
    const tableWrapperInstance: ClientSideTable<any, any> = tableWrapper
      .find(ClientSideTable)
      .instance() as ClientSideTable<any, any>;
    Reflect.set(tableWrapperInstance, 'refreshClientSideRowModelTimeout', 1);
    Reflect.set(
      tableWrapperInstance,
      'getTableApi',
      jest.fn(() => ({
        ensureIndexVisible: jest.fn(() => true),
        refreshInfiniteCache: jest.fn()
      }))
    );
    tableWrapper.update();
    tableWrapper.instance().forceUpdate();
    expect(setTimeoutSpy).toHaveBeenCalled();
    expect(clearTimeoutSpy).toHaveBeenCalled();
  });
  it('will not refresh client side row model the table API is unavailable', () => {
    const clearTimeoutSpy = jest.spyOn(global, 'clearTimeout');
    const setTimeoutSpy = jest.spyOn(global, 'setTimeout');
    const tableWrapper = Enzyme.mount(<ClientSideTable {...clientSideTableProps} />);
    const tableWrapperInstance: ClientSideTable<any, any> = tableWrapper
      .find(ClientSideTable)
      .instance() as ClientSideTable<any, any>;
    Reflect.set(
      tableWrapperInstance,
      'getTableApi',
      jest.fn(() => false)
    );
    tableWrapper.update();
    tableWrapper.instance().forceUpdate();
    // calls attributed to component lifecycle up to componentDidUpdate
    expect(setTimeoutSpy).toHaveBeenCalledTimes(4);
    expect(clearTimeoutSpy).not.toHaveBeenCalled();
  });
});
