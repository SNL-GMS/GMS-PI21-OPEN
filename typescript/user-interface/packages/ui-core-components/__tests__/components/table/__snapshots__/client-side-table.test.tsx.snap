// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Side Table Renders 1`] = `
<CoreTable
  columnDefs={
    Array [
      Object {
        "field": "test",
        "headerName": "test",
        "headerTooltip": "test",
        "width": 10,
      },
      Object {
        "field": "test2",
        "headerName": "test2",
        "headerTooltip": "test2",
        "sort": "asc",
        "width": 10,
      },
    ]
  }
  onCellClicked={[MockFunction]}
  onCellContextMenu={[MockFunction]}
  onGridReady={[Function]}
  onRowSelected={[MockFunction]}
  rowBuffer={10}
  rowData={
    Array [
      Object {
        "Ids": Array [
          1,
          2,
        ],
        "category": "normal",
        "color": "RED",
        "endTime": 2000,
        "first-in-table": true,
        "id": "1",
        "rationale": "user",
        "startTime": 1000,
        "type": "info",
        "versionId": 1,
      },
      Object {
        "Ids": Array [
          1,
          2,
        ],
        "category": "normal",
        "color": "GREEN",
        "endTime": 3000,
        "id": "2",
        "rationale": "user",
        "startTime": 2000,
        "type": "info",
        "versionId": 1,
      },
    ]
  }
/>
`;
